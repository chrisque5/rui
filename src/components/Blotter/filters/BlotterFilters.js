/* global BigInt */
import { useState, useEffect, useCallback } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import PropTypes from 'prop-types';
import {
  Form, Button, Row,
} from 'antd';
import moment from 'moment';
import { getSSESessionId, getResetBlotterSearch, getDateRangeLimit } from '../../../utils/selectors';
import { showErrorNotification } from '../../../utils/notifications';
import { loadBlotterDeals } from '../../../actions/dealActions';
import { resetBlotterSearch } from '../../../actions/uiActions';
import { loadBlotterSearchDateRangeLimit } from '../../../actions/systemSettingsActions';
import { ids, MAX_DEAL_ID_VAL } from '../../../utils/constants';

import DateField from './DateField';
import TextField from './TextField';
import DealId from './DealId';

const marginTop = 10;
const BlotterFilters = ({ closeDrawer }) => {
  const [form] = Form.useForm();
  const dispatch = useDispatch();
  const sessionId = useSelector(getSSESessionId);
  const MAX_DATE_RANGE = useSelector(getDateRangeLimit);
  const [dealIdCls, setDealIdCls] = useState(null);
  const [dateRangeCls, setDateRangeCls] = useState('highlight');
  const isBlotterSearchReset = useSelector(getResetBlotterSearch);
  useEffect(() => {
    dispatch(loadBlotterSearchDateRangeLimit());
  }, [dispatch]);

  const validateForm = useCallback(() => {
    let result = { valid: true };

    const fieldsToBeValidated = [ids.BROKER, ids.TRADER, ids.CUSTOMER];
    const dealId = form.getFieldValue([ids.DEAL_ID]);
    // If deal id has value, then date fields are not mandatory
    if (dealId) {
      fieldsToBeValidated.push(ids.DEAL_ID);
    } else {
      fieldsToBeValidated.push(ids.DATE_FROM);
      fieldsToBeValidated.push(ids.DATE_TO);
    }
    return form.validateFields(fieldsToBeValidated).then((values) => {
      const { dateFrom, dateTo } = values;
      if (moment(dateFrom) > moment(dateTo)) {
        result = {
          valid: false,
          error: 'From date cannot be greater than To date',
        };
      } else if (dateFrom && dateTo && dateTo.diff(dateFrom, 'days') > MAX_DATE_RANGE) {
        result = {
          valid: false,
          error: `Date range cannot be greater than ${MAX_DATE_RANGE} days`,
        };
      } else {
        result = {
          ...values,
          valid: true,
          subscribeToUpdates: !!sessionId,
          sessionId,
        };
      }
      return result;
    }).catch(() => {
      const dealIDError = form.getFieldError(ids.DEAL_ID);
      if (dealIDError.length > 0) {
        result = {
          valid: false,
          error: dealIDError.join('. '),
        };
      } else {
        result = {
          valid: false,
          error: 'Trade date fields are required for search',
        };
      }

      return result;
    });
  }, [form, sessionId, MAX_DATE_RANGE]);

  const handleSearch = useCallback((shouldCloseDrawer = true) => {
    validateForm().then((result) => {
      if (result.valid) {
        dispatch(loadBlotterDeals(result));
        if (shouldCloseDrawer) {
          closeDrawer();
        }
      } else {
        showErrorNotification('Validation Error', result.error);
      }
    });
  }, [dispatch, validateForm, closeDrawer]);

  const resetForm = useCallback(() => {
    form.resetFields();
    setDateRangeCls('highlight');
    setDealIdCls('');
    handleSearch(false);
  }, [form, handleSearch]);

  useEffect(() => {
    if (isBlotterSearchReset) {
      resetForm();
      dispatch(resetBlotterSearch(false));
    }
  }, [isBlotterSearchReset, dispatch, resetForm]);

  const onFormValuesChange = (changes) => {
    if (changes[ids.DEAL_ID]) {
      form.setFieldsValue({
        [ids.DATE_FROM]: null,
        [ids.DATE_TO]: null,
        [ids.BROKER]: null,
        [ids.CUSTOMER]: null,
        [ids.TRADER]: null,
      });
      const dealId = form.getFieldsValue(ids.DEAL_ID);
      if (dealId && dealId !== '') {
        setDealIdCls('highlight');
      }
      setDateRangeCls('');
    } else {
      form.resetFields([ids.DEAL_ID]);
      setDealIdCls('');
      setDateRangeCls('highlight');
    }
  };

  const dealIdValidator = async (rule, value) => {
    if (value || value === 0) {
      if (value.toString().indexOf('.') !== -1) {
        throw new Error('Decimals are not allowed in DealId');
      } else if (BigInt(value) > BigInt(MAX_DEAL_ID_VAL)) {
        throw new Error('Invalid ID value');
      } else if (value.toString().length > 19) {
        throw new Error('Maximum 19 digits allowed in DealId');
      } else if (value > 0) {
        return Promise.resolve();
      }
      throw new Error('DealId should be larger than 0');
    }
    return Promise.resolve();
  };

  const handleKeyDown = (event) => {
    if (event.key === 'Enter') {
      handleSearch();
    }
  };

  return (
    <div className="blotter-filter-container">
      <Form
        hideRequiredMark
        form={form}
        onValuesChange={onFormValuesChange}
      >
        <fieldset className={dealIdCls}>
          <legend>Search By Deal ID</legend>
          <Row type="flex">
            <DealId
              label="Deal ID"
              id={ids.DEAL_ID}
              testId={ids.DEAL_ID}
              validator={dealIdValidator}
              handleKeyDown={handleKeyDown}
              className="form-item-disabled"
            />
          </Row>
        </fieldset>
        <fieldset style={{ marginTop }} className={dateRangeCls}>
          <legend>Search By Trade Date</legend>
          <Form.Item label="Trade Date" className="blotter-trade-date" colon={false}>
            <DateField disabled id={ids.DATE_FROM} label="From" testId={ids.DATE_FROM} />
            <DateField id={ids.DATE_TO} label="To" testId={ids.DATE_TO} />
          </Form.Item>
          <Row type="flex">
            <TextField label="Broker" id={ids.BROKER} testId={ids.BROKER} handleKeyDown={handleKeyDown} />
          </Row>
          <Row type="flex">
            <TextField label="Customer" id={ids.CUSTOMER} testId={ids.CUSTOMER} handleKeyDown={handleKeyDown} />
          </Row>
          <Row type="flex">
            <TextField label="Trader" id={ids.TRADER} testId={ids.TRADER} handleKeyDown={handleKeyDown} />
          </Row>
        </fieldset>
        <Row type="flex" justify="end" className="action-buttons" style={{ marginTop }}>
          <Button
            style={{ marginRight: 3 }}
            data-testid="blotterReset"
            onClick={resetForm}
          >
            Reset
          </Button>
          <Button
            htmlType="button"
            onClick={handleSearch}
            type="primary"
            data-testid="blotterSearch"
            style={{ marginRight: 10 }}
          >
            Search
          </Button>
        </Row>
      </Form>
    </div>
  );
};

BlotterFilters.propTypes = {
  closeDrawer: PropTypes.func.isRequired,
};
export default BlotterFilters;
