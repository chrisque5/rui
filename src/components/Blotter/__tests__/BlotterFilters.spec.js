import {
  render, fireEvent, waitFor, getByTitle, getAllByTitle,
} from 'test-utils/testUtils';
import redux from 'react-redux';
import BlotterFilters from '../filters/BlotterFilters';

jest.mock('react-redux', () => ({
  useDispatch: () => jest.fn(),
  useSelector: () => jest.fn(),
}));
jest.mock('moment', () => () => jest.requireActual('moment')('2020-09-02T00:00:00.000Z'));

describe('<BlotterFilters/>', () => {
  const dispatchSpy = jest.spyOn(redux, 'useDispatch');
  const closeDrawer = jest.fn();
  afterEach(() => {
    dispatchSpy.mockClear();
  });
  const renderBloterFilters = () => render(
    <BlotterFilters closeDrawer={closeDrawer} />,
  );
  test('renders snapshot as expected', async () => {
    const { container } = await renderBloterFilters();
    expect(container).toMatchSnapshot();
  });

  test('verify search event', async () => {
    const { getByTestId } = await renderBloterFilters();
    const searchBtn = await waitFor(() => getByTestId('blotterSearch'));
    fireEvent.submit(searchBtn);
    expect(dispatchSpy).toHaveBeenCalled();
  });

  test('verify date change event', async () => {
    const { getByTestId } = await renderBloterFilters();

    const dateInput = await waitFor(() => getByTestId('dateFrom'));
    fireEvent.mouseDown(dateInput);

    const selectedDateElem = await waitFor(() => getByTitle(document.body, '2020-09-01'));
    fireEvent.click(selectedDateElem);

    expect(dateInput.value).toBe('09/01/2020');

    const dateToInput = await waitFor(() => getByTestId('dateTo'));
    fireEvent.mouseDown(dateToInput);

    const dates = await waitFor(() => getAllByTitle(document.body, '2020-09-02'));
    fireEvent.mouseDown(dates[0]);

    expect(dateToInput.value).toBe('09/02/2020');
  });
});
