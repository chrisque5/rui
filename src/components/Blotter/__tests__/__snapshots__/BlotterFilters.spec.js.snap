// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BlotterFilters/> renders snapshot as expected 1`] = `
<div>
  <div
    class="blotter-filter-container"
  >
    <form
      class="ant-form ant-form-horizontal ant-form-hide-required-mark"
    >
      <fieldset
        class=""
      >
        <legend>
          Search By Deal ID
        </legend>
        <div
          class="ant-row"
          type="flex"
        >
          <div
            class="ant-row ant-form-item dealid"
            style="margin-bottom: 5px;"
          >
            <div
              class="ant-col ant-form-item-label"
            >
              <label
                class=""
                for="dealId"
                title="Deal ID"
              >
                Deal ID
              </label>
            </div>
            <div
              class="ant-col ant-form-item-control"
            >
              <div
                class="ant-form-item-control-input"
              >
                <div
                  class="ant-form-item-control-input-content"
                >
                  <span
                    class="ant-input-affix-wrapper dealid"
                  >
                    <input
                      class="ant-input"
                      data-testid="dealId"
                      id="dealId"
                      maxlength="19"
                      type="number"
                      value=""
                    />
                    <span
                      class="ant-input-suffix"
                    >
                      <span
                        aria-label="close-circle"
                        class="anticon anticon-close-circle ant-input-clear-icon-hidden ant-input-clear-icon"
                        role="button"
                        tabindex="-1"
                      >
                        <svg
                          aria-hidden="true"
                          data-icon="close-circle"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="64 64 896 896"
                          width="1em"
                        >
                          <path
                            d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
                          />
                        </svg>
                      </span>
                    </span>
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </fieldset>
      <fieldset
        class="highlight"
        style="margin-top: 10px;"
      >
        <legend>
          Search By Trade Date
        </legend>
        <div
          class="ant-row ant-form-item blotter-trade-date"
        >
          <div
            class="ant-col ant-form-item-label"
          >
            <label
              class="ant-form-item-no-colon"
              title="Trade Date"
            >
              Trade Date
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <div
                  class="ant-row ant-form-item blotter-date-filter ant-form-item-is-validating"
                >
                  <div
                    class="ant-col ant-form-item-label"
                  >
                    <label
                      class="ant-form-item-required"
                      for="dateFrom"
                      title="From"
                    >
                      From
                    </label>
                  </div>
                  <div
                    class="ant-col ant-form-item-control"
                  >
                    <div
                      class="ant-form-item-control-input"
                    >
                      <div
                        class="ant-form-item-control-input-content"
                      >
                        <div
                          class="ant-picker"
                          style="width: 100%;"
                        >
                          <div
                            class="ant-picker-input"
                          >
                            <input
                              autocomplete="off"
                              data-testid="dateFrom"
                              id="dateFrom"
                              placeholder="Select date"
                              readonly=""
                              size="12"
                              title="09/02/2020"
                              value="09/02/2020"
                            />
                            <span
                              class="ant-picker-suffix"
                            />
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="ant-row ant-form-item blotter-date-filter ant-form-item-is-validating"
                >
                  <div
                    class="ant-col ant-form-item-label"
                  >
                    <label
                      class="ant-form-item-required"
                      for="dateTo"
                      title="To"
                    >
                      To
                    </label>
                  </div>
                  <div
                    class="ant-col ant-form-item-control"
                  >
                    <div
                      class="ant-form-item-control-input"
                    >
                      <div
                        class="ant-form-item-control-input-content"
                      >
                        <div
                          class="ant-picker"
                          style="width: 100%;"
                        >
                          <div
                            class="ant-picker-input"
                          >
                            <input
                              autocomplete="off"
                              data-testid="dateTo"
                              id="dateTo"
                              placeholder="Select date"
                              readonly=""
                              size="12"
                              title="09/02/2020"
                              value="09/02/2020"
                            />
                            <span
                              class="ant-picker-suffix"
                            />
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-row"
          type="flex"
        >
          <div
            class="ant-row ant-form-item"
            style="margin-bottom: 5px;"
          >
            <div
              class="ant-col ant-form-item-label"
            >
              <label
                class=""
                for="broker"
                title="Broker"
              >
                Broker
              </label>
            </div>
            <div
              class="ant-col ant-form-item-control"
            >
              <div
                class="ant-form-item-control-input"
              >
                <div
                  class="ant-form-item-control-input-content"
                >
                  <span
                    class="ant-input-affix-wrapper"
                  >
                    <input
                      class="ant-input"
                      data-testid="broker"
                      id="broker"
                      type="text"
                      value=""
                    />
                    <span
                      class="ant-input-suffix"
                    >
                      <span
                        aria-label="close-circle"
                        class="anticon anticon-close-circle ant-input-clear-icon-hidden ant-input-clear-icon"
                        role="button"
                        tabindex="-1"
                      >
                        <svg
                          aria-hidden="true"
                          data-icon="close-circle"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="64 64 896 896"
                          width="1em"
                        >
                          <path
                            d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
                          />
                        </svg>
                      </span>
                    </span>
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-row"
          type="flex"
        >
          <div
            class="ant-row ant-form-item"
            style="margin-bottom: 5px;"
          >
            <div
              class="ant-col ant-form-item-label"
            >
              <label
                class=""
                for="customer"
                title="Customer"
              >
                Customer
              </label>
            </div>
            <div
              class="ant-col ant-form-item-control"
            >
              <div
                class="ant-form-item-control-input"
              >
                <div
                  class="ant-form-item-control-input-content"
                >
                  <span
                    class="ant-input-affix-wrapper"
                  >
                    <input
                      class="ant-input"
                      data-testid="customer"
                      id="customer"
                      type="text"
                      value=""
                    />
                    <span
                      class="ant-input-suffix"
                    >
                      <span
                        aria-label="close-circle"
                        class="anticon anticon-close-circle ant-input-clear-icon-hidden ant-input-clear-icon"
                        role="button"
                        tabindex="-1"
                      >
                        <svg
                          aria-hidden="true"
                          data-icon="close-circle"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="64 64 896 896"
                          width="1em"
                        >
                          <path
                            d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
                          />
                        </svg>
                      </span>
                    </span>
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-row"
          type="flex"
        >
          <div
            class="ant-row ant-form-item"
            style="margin-bottom: 5px;"
          >
            <div
              class="ant-col ant-form-item-label"
            >
              <label
                class=""
                for="trader"
                title="Trader"
              >
                Trader
              </label>
            </div>
            <div
              class="ant-col ant-form-item-control"
            >
              <div
                class="ant-form-item-control-input"
              >
                <div
                  class="ant-form-item-control-input-content"
                >
                  <span
                    class="ant-input-affix-wrapper"
                  >
                    <input
                      class="ant-input"
                      data-testid="trader"
                      id="trader"
                      type="text"
                      value=""
                    />
                    <span
                      class="ant-input-suffix"
                    >
                      <span
                        aria-label="close-circle"
                        class="anticon anticon-close-circle ant-input-clear-icon-hidden ant-input-clear-icon"
                        role="button"
                        tabindex="-1"
                      >
                        <svg
                          aria-hidden="true"
                          data-icon="close-circle"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="64 64 896 896"
                          width="1em"
                        >
                          <path
                            d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
                          />
                        </svg>
                      </span>
                    </span>
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </fieldset>
      <div
        class="ant-row ant-row-end action-buttons"
        style="margin-top: 10px;"
        type="flex"
      >
        <button
          class="ant-btn"
          data-testid="blotterReset"
          style="margin-right: 3px;"
          type="button"
        >
          <span>
            Reset
          </span>
        </button>
        <button
          class="ant-btn ant-btn-primary"
          data-testid="blotterSearch"
          style="margin-right: 10px;"
          type="button"
        >
          <span>
            Search
          </span>
        </button>
      </div>
    </form>
  </div>
</div>
`;
